### ISSUE TITLE
[4.x]: Discount automatically applied to all orders when no coupon code is set

### ISSUE THREAD

**Issue by samueldraper** [craftcms/commerce#3132]
### Description

Users automatically seeing coupon codes applied at checkout, and discount applied for which they are not eligible.

Cause seems to be because no coupon code was specified therefore (one of several but not all) discounts were applied to cart.

### Steps to reproduce

1. Create discount rules
2. Do not add a coupon code

### Expected behavior

Should not apply to all carts.
Coupon code should be mandatory.

### Other info

Based on  #3131 this issue **is actually** expected behaviour. If that is the case a UX enhancement would be much needed in order to avoid the possibility of this happening. Even as simple as some validation: "Are you sure you don't want to add a coupon code? This will result in discount being applied to all carts that meet the criteria".

### Craft CMS version

4

### Craft Commerce version

4

### Installed plugins and versions

-

**Comment by lukeholder** [comment]
Coupon codes are just a condition for the discount, just like any other condition on the Discount. 
The feature is `Promotions > Discounts`, not `Promotions > Coupons`.

We could make it cleaner within the discount edit page, though. This is the first time someone has submitted this as a misunderstanding. I will leave the ticket open and see what I can do.

**Comment by soulseekah** [comment]
We've lost almost 1000 Euros due to this misunderstanding. A content manager deleted the coupon code but not the discount itself...

This "feature" is what Sales are supposed to do, it's highly counterintuitive and should probably be behavior that's off by default (on for existing Discounts).

+1 on at least making it clearer on the page that the discount is automatically applied.

The documentation does hint at this behavior, but does not state this specifically as potentially unwanted behavior.

P.S.: Such discounts actually stack.

**Comment by Romanavr** [comment]
@lukeholder 

This is completely unexpected behaviour, and I see no reason why the coupon code should be automatically applied when no coupon code is generated? Could you give me some real practical examples where it should be used?
I accidentally ran into the same issue and I still confused. We lost some money because of it.

We have sales for the _automatic_ discounts, and this is logically they can be applied to any customer. They are even located in a separate section. But I'll be honest, I don't understand at all why a coupon code works as a sale, especially when there is no actual coupon code lol.

Could you please reconsider this behaviour of coupon codes and ,maybe, mark it as a bug? Or perhaps I don't know the reason why it's done in such way.

Either way, thanks for any answer!

**Comment by nfourtythree** [comment]
Hi all

The functionality in Commerce for discounts and coupon codes has been the same since its inception. If there are no coupons set on the discount the user does not need to enter a discount to potentially be eligible. This is intended functionality as there are many types of discount that do not require a code to be entered, for example spend $100 and get 10% off or buy X number of items and get a discount. 

It is obviously disappointing in how things have turn out for your projects. So in a chance to improve the user experience and hopefully attempt to negate the issue of a store manager accidentally removing all coupons we are going to look to implement a new "Requires coupon code" light switch on the coupons tab. This will then have to be enabled to be able to start using the coupons but will also act as a "fail safe" that if the light switch is enabled but there are no coupons (maybe because the have been deleted legitimately or accidentally) we will know that the discount can not be matched to an order.

Thank you for your patience and feedback. 

We will keep this issue open and reference it when the new feature is available.

Thanks!

**Comment by Romanavr** [comment]
Thank you for that. We are really looking forward to this feature in the new release)

**Comment by nfourtythree** [comment]
Hey @soulseekah 

You mentioned in your comment that:

> A content manager deleted the coupon code but not the discount itself

When working on this [new feature](https://github.com/craftcms/commerce/pull/3660) we [came across a bug](https://github.com/craftcms/commerce/pull/3660/commits/7137c709c99b7e5413f8fb757f7c9d998c1b182d) in Commerce 4 that meant in the control panel it was impossible to remove all coupons from a discount. When on the discount's edit page if you remove all the coupon lines and save the discount the codes will not actually be deleted and will appear again when you view the discount.

I know time has passed since then but if you remember how that content manager did it, it could be useful for us to know so that we can cover all the scenarios.

In terms of the new feature, we are looking to get this included in versions `4.7` and `5.2` of Commerce.

Thanks!

**Comment by soulseekah** [comment]
@nfourtythree we are absolutely certain that we were able to a) create discounts without coupon codes (we have an internal video of this demonstrating the issue), b) remove a coupon code on an existing discount without any errors or them reappearing (this was demonstrated internally on a live video call with no recording unfortunately).

The version we are using is 4.6.0 (ref: 0384a483083a7793f734c43af050a0a85334c021).

**Comment by nfourtythree** [comment]
Hi @soulseekah 

Thank you for your response, we appreciate you taking the time to give us that information.

> we are absolutely certain that we were able to a) create discounts without coupon codes (we have an internal video of this demonstrating the issue)

This is correct, you will have been able to create discounts without coupon codes. This, as explained previously, is by design and has always been the case in Commerce. Giving the flexibility to create a variety of different ways of creating promotions for developers and store administrators.

>  b) remove a coupon code on an existing discount without any errors or them reappearing (this was demonstrated internally on a live video call with no recording unfortunately).
> The version we are using is 4.6.0

I have just checked out version `4.6.0` to see if the bug I mentioned (about not being able to fully remove all coupons codes) existed and can confirm it seems that it was present (video below). So, unfortunately, we can't quite see how the store admin in your project was able to delete/remove all the coupons for a discount.

https://github.com/user-attachments/assets/fafd5fdc-83a2-40cf-b5c2-2bcf471fb00e

If you do remember/figure out how they achieved that, please let us know. 

But for now we have [the PR](https://github.com/craftcms/commerce/pull/3660) ready for the new "Require Coupon Code" feature which should hopefully make things a lot clearer for store admins going forward.

We will update this issue when the new version and feature are released.

Thanks!

**Comment by Romanavr** [comment]
Hi @nfourtythree 

That's right, when the coupon table is completely empty, you can't save it. Well, you can, but it'll revert everything (before your fix).
However, you can **create** a discount **with an empty 'coupon codes' table**.

And in that case, it's a huge problem for us. I believe it should be solved with your [PR, required coupon code](https://github.com/craftcms/commerce/pull/3660).
BTW, when do you plan to release it?

Anyway, thanks for the reply.

**Comment by lukeholder** [comment]
Craft Commerce 5.2.0 and 4.7.0 are now out and contain a new "require a coupon code" checkbox on the discount that needs to be checked to be able to add discount codes to the discount. Hope that helps!

**Comment by Romanavr** [comment]
> Craft Commerce 5.2.0 and 4.7.0 are now out and contain a new "require a coupon code" checkbox on the discount that needs to be checked to be able to add discount codes to the discount. Hope that helps!

Sorry, but I think it's still not working correctly.
Let me know if we need to turn this into a discussion.

We have a project with customizable logic for gifts, where we generate gifts with their coupon codes.

Since the new attribute defaults to `false`, it caused the same problem as described here.
Our customers started receiving unexpected discounts, which made the situation even worse.

So, I probably don't understand the concept of discounts/sales.
We have **sales** that should work without coupon codes, and that's fine.
And we have **discounts** that can work in both ways, with and without coupon codes. Why does it work this way?

Can't we just have discount codes and Sales separately?
I mean, **in my picture**, **Sales** should work by conditions, like it's happening right now.
And **Discounts** should **ONLY apply when coupon codes are used** **and certain conditions are met**.

Basically, right now **Discount without coupon code = Sale**, no?

**Comment by lukeholder** [comment]
Sales (replaced with the pricing catalog in Commerce 5) are rules that apply to purchasables outside the cart. 

Discounts are rules that apply to the cart. One of those rules is a coupon code that can be added to the cart to apply the coupon.

That's how it's worked since the beginning of commerce with improvements along the way. 

Happy to hear more thoughts in a discussion thread.
